### PR TITLE
[server] tslint: no-floating-promises

### PR DESCRIPTION
**Pull request by geropl**
## Description
Introduces the tslint rule [no-floating-promises](https://palantir.github.io/tslint/rules/no-floating-promises/) to the `server` code base to avoid UnhandledPromisesRejections (at least in our code, that is).

It also ensures that `yarn lint` is run as part of `yarn build` and `leeway build`.

## Related Issue(s)

Fixes #

## Release Notes

```release-note
none
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **geropl** after the PR has been reviewed.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/ee/db-sync/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/db-sync/OWNERS)**
- **[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)**
- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/8012#issuecomment-1029119792" title="Approved">geropl</a>*

Associated issue requirement bypassed by: *<a href="https://github.com/gitpod-io/gitpod/pull/8012#issuecomment-1029119792" title="Approved">geropl</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/ee/db-sync/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/db-sync/OWNERS)~~ [geropl]
- ~~[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)~~ [geropl]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl]
- ~~[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)~~ [geropl]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/server/tslint.yaml
Comment (AlexTugarev) [inline components/server/tslint.yaml:7]: 🎉 thank you so much for starting this!

File: components/server/src/express/ws-handler.ts
Comment (jankeromnes) [inline components/server/src/express/ws-handler.ts:73]: Out of curiosity, why put the `.catch()` after the `.finally()` and not the other way around?

File: components/server/src/projects/projects-service.ts
Comment (jankeromnes) [inline components/server/src/projects/projects-service.ts:133]: Maybe log this?

```suggestion
        /** no await */ this.getProjectOverviewCached(installer, project)
            .catch(err => log.warn('could not pre-fetch project overview', err));
```

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [inline components/server/src/workspace/gitpod-server-impl.ts:463]: Oops 🤯

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [inline components/server/src/workspace/gitpod-server-impl.ts:1694]: Typo

```suggestion
            /** no await */ this.deleteProject(ctx, project.id)
```

Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts:1694]: Actually, shouldn't we `await` here? 🤔

```ts
        const teamProjects = await this.projectsService.getTeamProjects(teamId);
        await Promise.all(teamProjects.map(project => this.deleteProject(ctx, project.id));
```

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [inline components/server/src/workspace/gitpod-server-impl.ts:1702]: Same here, I think we should await:

```ts
        const teamMembers = await this.teamDB.findMembersByTeam(teamId);
        await Promise.all(teamMembers.map(member => this.removeTeamMember(ctx, teamId, member.userId)));
```

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (AlexTugarev) [reply components/server/src/workspace/gitpod-server-impl.ts:463]: 👀

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (AlexTugarev) [reply components/server/src/workspace/gitpod-server-impl.ts:1694]: don't think so. why blocking the deletion of a team?

**Review by jankeromnes — APPROVED** [review]
Agree with @AlexTugarev -- this is pure awesome ✨ (and I already know another rule I'd like to add here 😇)

Looks good but holding due to a few comments and also @AlexTugarev sorry for ~~stealing~~co-reviewing this, holding for your approval too.

/hold

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: df8a110c467c1b97523fdad434a82665b271dcfc</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/8012#issuecomment-1029119792" title="Approved">geropl</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/8012#pullrequestreview-873078169" title="Approved">jankeromnes</a>*

Associated issue requirement bypassed by: *<a href="https://github.com/gitpod-io/gitpod/pull/8012#issuecomment-1029119792" title="Approved">geropl</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/ee/db-sync/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/db-sync/OWNERS)~~ [geropl,jankeromnes]
- ~~[components/ee/payment-endpoint/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ee/payment-endpoint/OWNERS)~~ [geropl,jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl,jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl,jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl,jankeromnes]
- ~~[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)~~ [geropl,jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/src/workspace/gitpod-server-impl.ts:1694]: Because when you delete a team, you want to be sure everything is gone, right? I.e. if this breaks mid-way due to a bug, maybe it's preferable to get an error and reach out to support for full deletion, as opposed to leaving potentially sensitive data behind? 💭

But maybe I'm being too paranoid here and deleting async / in the background is actually fine. 😊

Comment (AlexTugarev) [reply components/server/src/workspace/gitpod-server-impl.ts:1694]: We had the very same situation for user accounts and this would block re-creating team as well.

**Comment by AlexTugarev** [comment]
/hold cancel

👍🏻 nice!

File: components/server/src/express/ws-handler.ts
Comment (geropl) [reply components/server/src/express/ws-handler.ts:73]: `finally` can throw (again), that's why the lint wants the `catch` to be the last combinator.

File: components/server/src/workspace/gitpod-server-impl.ts
Comment (geropl) [reply components/server/src/workspace/gitpod-server-impl.ts:1702]: Thought the same, but was unsure about the context, so left it as is. We can discuss and adjust separately if needed. :+1:

**Pull request merged**
